File: Práctica 1/S1/S1R/lib/CarreraMontana.rb

require_relative "Carrera"

module Ruby
	public
	class CarreraMontana < Carrera

    attr_accessor :bicis
    

		def clonar()
			return CarreraMontana.new(@numero)
		end

		public
		def initialize(numero)
      super(numero)
      @bicis = Array.new
    end
    
    public 
    def correr
      puts "Comienza la carrera de carretera con #{@bicis.length}"
      abandonos = (@bicis.length*0.2).to_i
      puts "El numero de abandonos es #{abandonos}"
      for i in (0..abandonos-1)
        caido=rand(1..@bicis.size)
        puts "El participante #{caido} se ha retirado de la carrera de montania"
        @bicis.delete_at(caido-1)
      end
      @bicis = @bicis.shuffle
      puts "La carrera queda con #{@bicis.length}"
    end
    
    public 
    def mostrarranking
      puts "La carrera de montania ha acabado, resultados:"
      tamanio=@bicis.size-1
      for i in 0..tamanio
        puesto=i+1
        puts "Carrera #{@codigo} Puesto #{puesto} -> Bicicleta #{(@bicis[i].id).to_s}"
      end
    end
    
	end
end

File: Práctica 1/S1/S1J/src/s1/carreraMontana.java
package s1;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Random;

public class carreraMontana extends Carrera {
    
    carreraMontana(int codigo){
        super(codigo);
    }
    
    public void run(){
        
        System.out.println("Comienza la carrera de Montaña");
        Collections.shuffle(this.bicicletas); //Desordenamos los participantes, para simular que están compitiendo
        
        int abandonos = (int) ((int)this.bicicletas.size() * 0.2);
        
        System.out.println("Se han retirado " + abandonos + " ciclistas en la carrera de Montaña con codigo " + this.codigoCarrera);
        
        for(int i=0;i<abandonos;i++){
            Random rand = new Random();
            int caido = rand.nextInt(this.bicicletas.size());
            System.out.println("El participante " + this.bicicletas.get(caido).getID() + " se ha retirado en la carrera de Montaña con codigo " + this.codigoCarrera);
            this.bicicletas.remove(caido);
        }
    }
    
    public void mostrarRanking(){
        System.out.println("\n La carrera de Montaña ha acabado, el resultado ha sido el siguiente: ");
        for(int i=0;i<this.bicicletas.size();i++){
            int puesto = i+1;
            System.out.println("Carrera " + this.codigoCarrera + ".Puesto " + puesto + ".- Bicicleta " + this.bicicletas.get(i).getID());
        }
    }
}